stonesoup/models/measurement/clutter.py

```diff
@@ -0,0 +1,191 @@
+
+        # Check that the chosen distribution matches the parameters given
+        if self.distribution == Distribution.uniform and self.state_space is None:
+            raise ValueError("To use the uniform distribution, the state_space parameter must be defined")
+        elif self.distribution == Distribution.normal and self.mean is None and self.covariance is None:
+            raise ValueError("To use the normal distribution, the mean and covariance parameters must be defined")
+
+    def function(self, ground_truths: Set[GroundTruthState], **kwargs) -> Set[Clutter]:
+        """
+        Use the defined distribution and parameters to create simulated clutter
+        for the current time step. Return this clutter to the calling sensor so
+        that it can be added to the measurements.
+
+        Parameters
+        ----------
+        ground_truths : a set of :class:`~.GroundTruthState`
+            The truth states which exist at this time step.
+        
+        Returns
+        -------
+        : set of :class:`~.Clutter`
+            The simulated clutter.
+        """
+        # Extract the timestamp from the ground_truths. Groundtruth is
+        # necessary to get the proper timestamp. If there is no
+        # groundtruth return a set of no Clutter.
+        if not ground_truths:
+            return set()
+
+        # Generate the clutter for this time step
+        timestamp = next(iter(ground_truths)).timestamp
+        clutter = set()
+        for _ in range(np.random.poisson(self.clutter_rate)):
+            if self.distribution == Distribution.uniform:
+                random_vector = self.generate_uniform_vector()
+            else:
+                random_vector = self.generate_normal_vector()
+            
+            # Make a State object with the random vector
+            state = State([0.0] * self.measurement_model.ndim_state, timestamp=timestamp)
+            state.state_vector[self.measurement_model.mapping, 0] += random_vector 
+
+            # Use the sensor's measurement model to incorporate the
+            # translation offset and sensor rotation. This will also
+            # convert the vector to the proper measurement space
+            # (polar coordinates). 
+            clutter_vector = self.measurement_model.function(state)
+
+            # Create a clutter object.
+            clutter.add(Clutter(state_vector=clutter_vector, 
+                                timestamp=timestamp,
+                                measurement_model=self.measurement_model))
+
+        return clutter
+    
+    def generate_uniform_vector(self):
+        """Generate a random vector in the uniform distribution according
+        to the model's defined state space.
+
+        Returns
+        -------
+        : :class:`numpy.array`
+            A uniform random vector with length equal to self.ndim_meas.
+        """
+        arr = np.array([np.random.default_rng().uniform(*space) 
```

Comment:
I was thinking that the way I implemented it made it easier on the user as it would save them the trouble of researching random generators and deciding whether to use numpy or scipy etc. But I do see the benefits of your suggestion. It would definitely make it more versatile to different distributions and remove the need for logic and checks. If someone else comments in the next day or so with a reason to keep it the way it is then I will keep it. Otherwise I'll change it!

Reply:
Hi. I started to make this change yesterday but I ran into a problem. `np.random.default_rng().multivariate_normal(some_mean, some_covariance)` is a numpy array. It is not a function, meaning that if it is passed as a distribution then there is actually nothing to call to get more random vectors. This would happen with all such distribution functions I believe. If there is some other way to pass in a function with the parameters already set please let me know! :)

One way around this would be to make a function like the one below and pass that in
```
def distribution(mean, covariance):  
    return np.random.default_rng().multivariate_normal(mean, covariance)
```

Or I could have two parameters in the `ClutterModel` called `distribution` and `dist_params` . Then the user could set 
```
distribution = np.random.default_rng().multivariate_normal
dist_params = [mean, covariance]
```
and then the `ClutterModel` would say 
`clutter_vector = self.distribution(*self.dist_params)`

Is there a preferred way to do this? Thanks for your help :)

Reply:
I suppose it gets difficult where you are calling the rng several times for the uniform distribution, with varying args, but once for the normal distribution.

What you've suggested sounds good. i.e. you have properties:
    * `distribution` where user could pass `np.random.default_rng().multivariate_normal` or `np.random.default_rng().uniform` etc.
    * `dist_params` where user could pass `(mean, covariance)` or `[(mean, cov), (mean2, cov2)]` or `space` or `[space, space2, space3]`

You could then do:
random_vector = np.array([dist_method(*arg) for arg in dist_args])
(as you would still need to make sure you get an array back if you are working with multiple spaces with uniform dist)
Though this is starting to look confusing.

I'm maybe being too picky here. The current implementation you have is fine. Just wondered if there was 1. a way, 2. need for, genericising the distribution.

Reply:
Ok thanks, I'll incorporate this into the code. The more I think about it the more I agree that a generic version would be good. It means we don't have to do all the type checking, and leaves it up to the user. I'll add a default uniform distribution to make it easier for people who don't want to do something fancy or want to use it quickly.

Reply:
Update: this idea has been implemented. I also updated the example gists so that they run with the newly defined model.